uint32 node_id
float64[] bow
float64[] bowp